integration_test/third_party_apps_test.go

```diff
@@ -237,6 +314,137 @@ func runLoggingTestCases(ctx context.Context, logger *logging.DirectoryLogger, v
 	return err
 }
 
+func runMetricsTestCases(ctx context.Context, logger *logging.DirectoryLogger, vm *gce.VM, testCaseBytes []byte) error {
+	var metrics []expectedMetric
+	var err error
+	if err = yaml.UnmarshalStrict(testCaseBytes, &metrics); err != nil {
+		return fmt.Errorf("could not unmarshal contents of expected_metrics.yaml: %v", err)
+	}
+	if err = validateMetrics(metrics); err != nil {
+		return fmt.Errorf("expected_metrics.yaml failed validation: %v", err)
+	}
+	logger.ToMainLog().Printf("Parsed expected_metrics.yaml: %+v", metrics)
+	// Wait for the representative metric first, which is intended to *always*
+	// be sent. If it doesn't exist, we fail fast and skip running the other metrics;
+	// if it does exist, we go on to the other metrics in parallel, by which point they
+	// have gotten a head start and should end up needing fewer API calls before being found.
+	// In both cases we make significantly fewer API calls which helps us stay under quota.
+	for _, metric := range metrics {
+		if !metric.Representative {
+			continue
+		}
+		err = assertMetric(ctx, logger, vm, metric)
+		if gce.IsExhaustedRetriesMetricError(err) {
+			return fmt.Errorf("representative metric %s not found, skipping remaining metrics", metric.Type)
+		}
+		break
+	}
+	// Give some catch-up time to the remaining metrics, which tend to be configured
+	// for a 60-second interval and sometimes require two data points.
+	logger.ToMainLog().Println("Found representative metric, sleeping before checking remaining metrics")
+	time.Sleep(120 * time.Second)
+	c := make(chan error, len(metrics))
+	for _, metric := range metrics {
+		if metric.Representative {
+			c <- nil
+			continue
+		}
+		metric := metric
```

Comment:
The `c <- nil` or the `metric := metric`?

I chose `c <- nil` as a simple way to prevent the `<-c` loop later on from hanging. We create a channel with a fixed number of elements that closes automatically after they've all been processed; if we skip the `err` from the representative metric then that channel will never be filled and  the `<-c` loop at the end goes forever. This might be worth a comment though. There are other ways to close channels gracefully that's not tied to the number of elements but IMO it's at least as complicated as this. Open to opinions on this.

I need `metric := metric` to capture it into a loop-local variable for the `go func`.

Reply:
`I need metric := metric to capture it into a loop-local variable for the go func.` 👍

Reply:
[Optional] FWIW, we normally add a comment that links to https://golang.org/doc/faq#closures_and_goroutines …

Reply:
You can actually just put the comment on the same line — that way there's less possibility that it'd get separated.